File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;

        generate(0, 0, n, "", res);

        return res;
    }

private:
    void generate(int o, int c, int n, string s, vector<string>& res) {
        if(o == n && c == n) {
            res.push_back(s);
            return;
        }
        if(o < n) {
            generate(o+1, c, n, s + '(', res);
        }

        if(o > c) {
            generate(o, c+1, n, s + ')', res);
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def generate(p, o, c):
            if c == o == n:
                res.append(p)
                return

            if c < o:
                generate(p + ')', o, c+1)
            
            if o < n:
                generate(p + '(', o + 1, c)
        
        generate("(", 1, 0)

        return res